libcontainer/container_linux.go

```diff
@@ -1235,11 +1235,39 @@ func (c *linuxContainer) prepareCriuRestoreMounts(mounts []*configs.Mount) error
 	// Now go through all mounts and create the mountpoints
 	// if the mountpoints are not on a tmpfs, as CRIU will
 	// restore the complete tmpfs content from its checkpoint.
+	umounts := []string{}
+	defer func() {
+		for _, u := range umounts {
+			if e := unix.Unmount(u, unix.MNT_DETACH); e != nil {
+				if e != unix.EINVAL {
+					// Ignore EINVAL as it means 'EINVAL target is not a mount point.'
+					// It probably has already been unmounted.
+					logrus.Warnf("Error during cleanup unmounting of %q (%v)\n", u, e)
+				}
+			}
+		}
+	}()
 	for _, m := range mounts {
 		if !isPathInPrefixList(m.Destination, tmpfs) {
 			if err := c.makeCriuRestoreMountpoints(m); err != nil {
 				return err
 			}
+			// If the mount point is a bind mount, we need to mount
+			// it now so that runc can create the necessary mount
+			// points for mounts in bind mounts.
+			// This also happens during initial container creation.
+			// Without this CRIU restore will fail
+			// See: https://github.com/opencontainers/runc/issues/2748
+			// It is also not necessary to order the mount points
+			// because during initial container creation mounts are
+			// set up in the order they are configured.
```

Comment:
Just curious -- what happens if the order of (nested) mounts in config is wrong? Will runc run fail?

I wrote a quick test and it works for me:
```bats
@test "runc run [wrong order of tmpfs mounts]" {
        update_config '   .mounts += [{
                                        type: "tmpfs",
                                        source: "tmpfs",
                                        destination: "/mount/sub/dir",
                                        options: ["ro", "nodev", "nosuid"]
                                },
                                {
                                        type: "tmpfs",
                                        source: "tmpfs",
                                        destination: "/mount",
                                        options: ["rw", "nodev", "nosuid"]
                                }]
                        | .process.args |= ["true"]'
        runc run test_busybox
        [ "$status" -eq 0 ]
}
```

It works no matter what the order of added mounts is. Same with bind mounts:
```bats
@test "runc run [wrong order of bind mounts]" {
        mkdir bind1 bind2
        update_config '   .mounts += [{
                                        source: "bind2",
                                        destination: "/mount/sub/dir",
                                        options: ["bind"]
                                },
                                {
                                        source: "bind1",
                                        destination: "/mount",
                                        options: ["bind"]
                                }]
                        | .process.args |= ["true"]'
        runc run test_busybox
        [ "$status" -eq 0 ]
}
```
(the test succeeds).

Reply:
No, `runc` will not fail.

In your first case the second tmpfs mount will just overmount the first tmpfs mount. Same for your bind mount test. The mountpoints for the first mounts will always be created, but they will not be visible in the container. You should see a failure if you do a `ls /mount/sub/dir` in `.process.args`.

I was first looking at the code base for some time to find the function that orders the mounts to be able to re-use that code. But, from what I saw, `runc` just goes through the mounts as they appear in `config.json`.